tests/test_restrictions_emmo.py

```diff
@@ -0,0 +1,160 @@
+"""Test the API with the EMMO ontology."""
+
+import unittest2 as unittest
+import rdflib
+import itertools
+from osp.core.ontology.entity import OntologyEntity
+from osp.core.ontology.relationship import OntologyRelationship
+from osp.core.ontology.attribute import OntologyAttribute
+from osp.core.ontology.oclass_restriction import Restriction, QUANTIFIER, RTYPE
+from osp.core.ontology.oclass_composition import Composition
+
+
+try:
+    from osp.core.namespaces import math
+except ImportError:  # When the EMMO ontology is not installed.
+    from osp.core.ontology import Parser
+    from osp.core.namespaces import _namespace_registry
+    Parser(_namespace_registry._graph).parse("emmo")
+    _namespace_registry.update_namespaces()
+    math = _namespace_registry.math
+
+
+class TestRestrictionsEmmo(unittest.TestCase):
+    """Test the restrictions in EMMO ontology."""
+
+    emmo_datatypes = (math.Integer, math.Real, math.Boolean, math.Vector,
+                      math.Matrix)
+
+    quantifiers = {rdflib.OWL.someValuesFrom: QUANTIFIER.SOME,
+                   rdflib.OWL.allValuesFrom: QUANTIFIER.ONLY,
+                   rdflib.OWL.cardinality: QUANTIFIER.EXACTLY,
+                   rdflib.OWL.minCardinality: QUANTIFIER.MIN,
+                   rdflib.OWL.maxCardinality: QUANTIFIER.MAX}
+
+    rtypes = {OntologyRelationship: RTYPE.RELATIONSHIP_RESTRICTION,
+              OntologyAttribute: RTYPE.ATTRIBUTE_RESTRICTION}
+
+    def iter_restrictions(self):
+        """Returns all the datatypes of the test wrapped in an iterator."""
+        restrictions = iter([])
+        for datatype in self.emmo_datatypes:
+            restrictions = itertools.chain(restrictions,
+                                           (r for r in datatype.axioms
+                                            if type(r) is Restriction))
+        return restrictions
+
+    def test___str__(self):
+        """Test the __str__ method."""
+        for r in self.iter_restrictions():
+            self.assertIs(type(r.__str__()), str)
+
+    def test_quantifier(self):
+        """Test the quantifier method.
+
+        Also tests _compute_target and _check_quantifier implicitly.
+        """
+        for r in self.iter_restrictions():
+            b_node = r._bnode
+            graph = r._graph
+
+            restriction_predicates = graph.predicates(subject=b_node)
+            quantifiers = [self.quantifiers[p] for p in restriction_predicates
+                           if p in self.quantifiers]
+            self.assertTrue(len(quantifiers) == 1, "The restriction has not "
+                                                   "exactly one quantifier.")
+            quantifier = quantifiers[0]
+            self.assertIs(quantifier, r.quantifier)
```

Comment:
I very much like how these tests are very general and therefore very many restriction objects are covered. I am wondering if there should be some test that checks only one or two specific restrictions. See also the comment for the __str__ test.

Reply:
The main difficulty here is probably how to get one specific restriction objects, since the result of ontology_class.axioms is in no specific order.

Reply:
I added the class TestSpecificRestrictionsEmmo for this purpose. As you say, getting one specific restriction object is tricky. 

Therefore I opted of creating the restriction, namely a blank node and the associated triples, and then instantiating a restriction object from the blank node. The information used to create the blank node is compared against the one yielded by the restriction object when the attributes are queried.